### ISSUE TITLE
MobileNet 在 1477823 后 loss 和 accuracy 均不下降

### ISSUE THREAD

**Issue by admk** [Aaron-Zhao123/mayo-dev#5]
step        | epoch   | top1       | top5       | loss
--------  | ------- | -------- | -------- | -----
799998  |              | 54.24% | 78.15%   |
1099998 |             | 55.17%   | 78.69% |
1249558 |             | 56.24% | 79.58% |
1477823 | 73.82   |  56.60% | 79.80%  | ~2.19
1514653 |              | 56.49% | 79.56%  | ~2.19
1680122 |              | 56.50% | 79.57%  | ~2.18

现在以 1477823 （epoch 73.82）为 baseline，learning rate 再乘 0.16 继续 train 观察效果

**Comment by admk** [comment]
`⠏ epoch: 76.09 | loss: 2.07    ± 26% | acc:  53.9% | ckpt: 76.07 | tp:  275/s`

**Comment by admk** [comment]
答案揭晓
---

傻逼了。部分conv 512的stride错误。现在已经改对了。前后对比：
```
+---------+--------------------+--------------------+
|         | Before             | After              |
+---------+--------------------+--------------------+
| Layer   | Shape              | Shape              |
+---------+--------------------+--------------------+
| images  | ? x 224 x 224 x 3  | ? x 224 x 224 x 3  |
| labels  | ? x 1001           | ? x 1001           |
| conv0   | ? x 112 x 112 x 32 | ? x 112 x 112 x 32 |
| conv1   | ? x 112 x 112 x 64 | ? x 112 x 112 x 64 |
| conv2   | ? x 56 x 56 x 128  | ? x 56 x 56 x 128  |
| conv3   | ? x 56 x 56 x 128  | ? x 56 x 56 x 128  |
| conv4   | ? x 28 x 28 x 256  | ? x 28 x 28 x 256  |
| conv5   | ? x 28 x 28 x 256  | ? x 28 x 28 x 256  |
| conv6   | ? x 14 x 14 x 512  | ? x 14 x 14 x 512  |
| conv7   | ? x 14 x 14 x 512  | ? x 14 x 14 x 512  |
| conv8   | ? x 7 x 7 x 512    | ? x 14 x 14 x 512  |
| conv9   | ? x 7 x 7 x 512    | ? x 14 x 14 x 512  |
| conv10  | ? x 4 x 4 x 512    | ? x 14 x 14 x 512  |
| conv11  | ? x 4 x 4 x 512    | ? x 14 x 14 x 512  |
| conv12  | ? x 2 x 2 x 1024   | ? x 7 x 7 x 1024   |
| conv13  | ? x 2 x 2 x 1024   | ? x 7 x 7 x 1024   |
| pool    | ? x 1 x 1 x 1024   | ? x 1 x 1 x 1024   |
| dropout | ? x 1 x 1 x 1024   | ? x 1 x 1 x 1024   |
| fc      | ? x 1 x 1 x 1001   | ? x 1 x 1 x 1001   |
| logits  | ? x 1001           | ? x 1001           |
+---------+--------------------+--------------------+
```
因为继续train效果不太理想，也无法正确反应train的epoch数，故决定重新开始做training。观察到初始loss也是6.9几开始，std为0%，属正常现象。0.14个epoch后：
```
⠼ epoch: 0.14 | loss: 6.33    ±  4% | acc:   1.1% | ckpt: 0.00 | tp:  250/s
```

**Comment by admk** [comment]
stride 修复于 91363bc

**Comment by admk** [comment]
现使用配置导出
---
```yaml
---
dataset:
    background_class: {has: true, use: true}
    channel_means: [0.485019607843137, 0.4579607843, 0.407607843137255]
    name: imagenet
    num_classes: 1001
    num_examples_per_epoch: {train: 1281167, validate: 50000}
    path: {train: imagenet/train-*, validate: imagenet/validation-*}
    shape: {channels: 3, height: 224, width: 224}
depth_multiplier: 1
hyper:
- &id002 {scale: 4.0e-05, type: tensorflow.contrib.layers.l2_regularizer}
-   kernel_size: [3, 3]
    normalizer_fn: &id001 {center: true, decay: 0.9997, epsilon: 0.001,
        scale: true, type: tensorflow.contrib.slim.batch_norm}
    padding: same
    type: convolution
    weights_initializer: &id003 {stddev: 0.09, type: tensorflow.truncated_normal_initializer}
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    normalizer_fn: *id001
    padding: same
    pointwise_regularizer: *id002
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   kernel_size: [7, 7]
    padding: valid
    stride: 2
    type: average_pool
- {keep_prob: 0.999, type: dropout}
-   activation_fn: null
    kernel_size: [1, 1]
    normalizer_fn: null
    num_outputs: num_classes
    type: convolution
    weights_regularizer: *id002
logits: logits
name: mobilenet_v1
net:
-   kernel_size: [3, 3]
    name: conv0
    normalizer_fn: *id001
    num_outputs: 32
    padding: same
    stride: 2
    type: convolution
    weights_initializer: *id003
    weights_regularizer: *id002
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv1
    normalizer_fn: *id001
    num_outputs: 64
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv2
    normalizer_fn: *id001
    num_outputs: 128
    padding: same
    pointwise_regularizer: *id002
    stride: 2
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv3
    normalizer_fn: *id001
    num_outputs: 128
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv4
    normalizer_fn: *id001
    num_outputs: 256
    padding: same
    pointwise_regularizer: *id002
    stride: 2
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv5
    normalizer_fn: *id001
    num_outputs: 256
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv6
    normalizer_fn: *id001
    num_outputs: 512
    padding: same
    pointwise_regularizer: *id002
    stride: 2
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv7
    normalizer_fn: *id001
    num_outputs: 512
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv8
    normalizer_fn: *id001
    num_outputs: 512
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv9
    normalizer_fn: *id001
    num_outputs: 512
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv10
    normalizer_fn: *id001
    num_outputs: 512
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv11
    normalizer_fn: *id001
    num_outputs: 512
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv12
    normalizer_fn: *id001
    num_outputs: 1024
    padding: same
    pointwise_regularizer: *id002
    stride: 2
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   depth_multiplier: 1
    depthwise_regularizer: null
    kernel_size: [3, 3]
    name: conv13
    normalizer_fn: *id001
    num_outputs: 1024
    padding: same
    pointwise_regularizer: *id002
    stride: 1
    type: depthwise_separable_convolution
    weights_initializer: *id003
-   kernel_size: [7, 7]
    name: pool
    padding: valid
    stride: 2
    type: average_pool
- {keep_prob: 0.999, name: dropout, type: dropout}
-   activation_fn: null
    kernel_size: [1, 1]
    name: fc
    normalizer_fn: null
    num_outputs: num_classes
    type: convolution
    weights_regularizer: *id002
-   axis: [1, 2]
    name: logits
    type: squeeze
preprocess:
    final:
    - {scale: 2.0, shift: -1.0, type: linear_map}
    train:
    - {type: distort_bbox}
    - {type: distort_color}
    - {type: random_flip}
    validate:
    - {central_fraction: 0.875, type: central_crop}
system:
    batch_size: 256
    checkpoint: {load: latest, save: true}
    log_level: {mayo: info, tensorflow: 2}
    max_steps: 10000000
    num_gpus: 4
    num_preprocess_threads: 16
    search_paths:
        checkpoints: [checkpoints]
        datasets: [datasets]
        summaries: [summaries]
    use_pdb: true
train:
    learning_rate: {_num_epochs_per_decay: 30.0, decay_rate: 0.16, decay_steps: 150136.7578125,
        learning_rate: 0.1, staircase: true, type: tensorflow.train.exponential_decay}
    moving_average_decay: 0.9999
    optimizer: {decay: 0.9, epsilon: 1.0, momentum: 0.9, type: tensorflow.train.RMSPropOptimizer}
```

进度更新
---
```
⠦ epoch: 0.21 | loss: 6.3     ±  2% | acc:   1.4% | ckpt: 0.00 | tp:  278/s
⠼ epoch: 0.24 | loss: 6.17    ±  2% | acc:   1.9% | ckpt: 0.00 | tp:  280/s
```

**Comment by admk** [comment]
Validation accuracy 还是无法达到 [baseline 70.7%/89.5%](https://github.com/tensorflow/models/tree/master/slim)，及 [zehaos 的版本 66.51%/87.09%](https://github.com/Zehaos/MobileNet)。
```
⣿ [100.00%] top1: 65.72%, top5: 86.44% (362.9 imgs/sec)
```

**Comment by admk** [comment]
step   | top 1  | top 5
---    | ---    | ---
625847 | 65.81% | 86.48%
595847 | 65.86% | 86.49%
525848 | 65.74% | 86.43%
455446 | 65.71% | 86.43%

**Comment by admk** [comment]
Using pre-trained model from tensorflow/models/slim, our evaluation produces top1: 55.92%, top5: 79.12%, which is a lot worse than the baseline 70.7%/89.5%. A quick comparison between these two did not find differences between them in preprocessing steps and the model.

I suspect the reason for discrepancy lies in `eval.py`, with exponential moving average as the most probable cause, will look further into this.

**Comment by admk** [comment]
The reason for the above-mentioned evaluation discrepancy was found in the choice of the activation function. The [original mobilenet](https://github.com/tensorflow/models/blob/master/slim/nets/mobilenet_v1.py) uses `tf.nn.relu6`, whereas we use the default activation `tf.nn.relu` in `models/mobilenet.yaml`. Fixing this resulted in accuracy which equals to the baseline.

Unfortunately this most likely doesn't explain why we cannot train mobilenet to the baseline accuracy, but we have strong reasons to believe the problem lies within `mayo/train.py` and/or `trainers/mobilenet.yaml`. I will consider closing this issue because I give up on training again from scratch.

**Comment by admk** [comment]
Evaluated all checkpoints.

step   | top1  | top5
---    | ---   | ---
495446 | 65.76 | 86.45
500446 | 65.72 | 86.44
505446 | 65.8  | 86.46
510446 | 65.81 | 86.48
515446 | 65.75 | 86.41
520446 | 65.73 | 86.45
525446 | 65.71 | 86.46
525848 | 65.74 | 86.43
530847 | 65.79 | 86.44
535847 | 65.81 | 86.47
540847 | 65.81 | 86.5
545847 | 65.8  | 86.44
550847 | 65.78 | 86.49
555847 | 65.76 | 86.49
560847 | 65.75 | 86.44
565847 | 65.77 | 86.48
570847 | 65.79 | 86.48
575847 | 65.76 | 86.48
580847 | 65.8  | 86.43
585847 | 65.78 | 86.44
590847 | 65.76 | 86.45
595847 | 65.85 | 86.47
600847 | 65.79 | 86.49
605847 | 65.81 | 86.48
610847 | 65.79 | 86.47
615847 | 65.8  | 86.48
620847 | 65.81 | 86.5
625847 | 65.81 | 86.49
